pkg/cmd/provisioning/alibabacloud/delete-ram-users.go

```diff
@@ -0,0 +1,123 @@
+	DeleteRAMUsersOpts = options{
+		Region:         "",
+		Name:           "",
+		CredRequestDir: "",
+	}
+)
+
+//detachComponentPolicy detach the specific ram policy from the ram user
+func detachComponentPolicy(client alibabacloud.Client, policyName, userName string) error {
+	req := ram.CreateDetachPolicyFromUserRequest()
+	req.PolicyName = policyName
+	req.PolicyType = ramPolicyType
+	req.UserName = userName
+	_, err := client.DetachPolicyFromUser(req)
+	return err
+}
+
+//deleteComponentPolicy delete the specific ram policy
+func deleteComponentPolicy(client alibabacloud.Client, policyName string) error {
+	req := ram.CreateDeletePolicyRequest()
+	req.PolicyName = policyName
+	_, err := client.DeletePolicy(req)
+	return err
+}
+
+//deleteComponentUser delete the specific component ram user
+func deleteComponentUser(client alibabacloud.Client, userName string) error {
+	req := ram.CreateDeleteUserRequest()
+	req.UserName = userName
+	_, err := client.DeleteUser(req)
+	return err
+}
+
+func deleteRAMUsers(client alibabacloud.Client, name, credReqDir string) error {
+	// Process directory
+	credRequests, err := provisioning.GetListOfCredentialsRequests(credReqDir)
+	if err != nil {
+		return errors.Wrap(err, "Failed to process files containing CredentialsRequests")
+	}
+
+	for _, credReq := range credRequests {
+		userName, _ := generateRAMUserName(fmt.Sprintf("%s-%s-%s", name, credReq.Namespace, credReq.Name))
+		listPoliciesReq := ram.CreateListPoliciesForUserRequest()
+		listPoliciesReq.UserName = userName
+		listPoliciesRes, err := client.ListPoliciesForUser(listPoliciesReq)
+		if err != nil {
+			return errors.Wrap(err, "Failed to list ram policies for component user")
+		}
+		//detach each policy from user
+		for _, userPolicy := range listPoliciesRes.Policies.Policy {
+			//detach component policy from the existing ram user
+			err := detachComponentPolicy(client, userPolicy.PolicyName, userName)
+			if err != nil {
+				return errors.Wrap(err, "Failed to detach ram policy from user")
+			}
+			//delete component ram policy
+			err = deleteComponentPolicy(client, userPolicy.PolicyName)
+			if err != nil {
+				return errors.Wrap(err, "Failed to delete component ram policy")
+			}
+		}
+		//delete component ram user
+		err = deleteComponentUser(client, userName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to delete component user")
+		}
+	}
+	return nil
+}
+
+func deleteRAMUsersCmd(cmd *cobra.Command, args []string) {
+	envProvider := provider.NewEnvProvider()
+	profileProvider := provider.NewProfileProvider()
+	pc := provider.NewProviderChain([]provider.Provider{envProvider, profileProvider})
+	cred, err := pc.Resolve()
+	if err != nil {
+		log.Fatalf("Failed to create a client: %v", err)
+	}
+	clientConfig := sdk.NewConfig().WithScheme(apiRequestScheme)
+	client, err := alibabacloud.NewClient(DeleteRAMUsersOpts.Region, clientConfig, cred)
+	if err != nil {
+		log.Fatalf("Failed to create a client: %v", err)
+	}
```

Comment:
Since this piece of code is repeated in all the subcommands, can we share this logic in pkg/alibabacloud/client.go for created a "default" client that sets up this common provider chain?

Reply:
good idea! have updated in the new commit